pos_barcode_tare/static/src/js/pos_barcode_tare.js

```diff
@@ -0,0 +1,359 @@
+        }
+
+        return result || 0;
+    };
+
+    // This configures read action for tare barcode. A tare barcode contains a
+    // fake product ID and the weight to be subtracted from the product in the
+    // latest order line.
+    screens.ScreenWidget.include(
+        {
+            barcode_tare_action: function (code) {
+                try {
+                    var order = this.pos.get_order();
+                    var last_order_line = order.get_last_orderline();
+                    var tare_weight = code.value;
+                    last_order_line.set_tare(tare_weight);
+                } catch (error) {
+                    var title = _t("We can not apply this tare barcode.");
+                    var popup = {title: title, body: error.message};
+                    this.gui.show_popup('error', popup);
+                }
+            },
+            // Setup the callback action for the "weight" barcodes.
+            show: function () {
+                this._super();
+                this.pos.barcode_reader.set_action_callback(
+                    'tare',
+                    _.bind(this.barcode_tare_action, this));
+            },
+        });
+
+    // This create a new button on top of action widget. This button links to
+    // the barcode label printing screen defined below.
+    var TareScreenButton = screens.ActionButtonWidget.extend({
+        template: 'TareScreenButton',
+
+        button_click: function () {
+            this.gui.show_screen('tare');
+        },
+    });
+
+    screens.define_action_button({
+        'name': 'tareScreenButton',
+        'widget': TareScreenButton,
+    });
+
+    // This is a new screen that reads weight from the electronic scale and
+    // create a barcode label encoding the weight. The screen shows a preview
+    // of the label. The user is expected to check if the preview matches what's
+    // measured on the scale. The barcode image is generated by the report
+    // module.
+    var TareScreenWidget = screens.ScreenWidget.extend({
+        template: 'TareScreenWidget',
+        next_screen: 'products',
+        previous_screen: 'products',
+        default_tare_value: 0.0,
+        weight_barcode_prefix: null,
+
+        show: function () {
+            this._super();
+            // Fetch the unit of measure used to save the tare
+            this.kg_unit = get_unit(this.pos, "kg");
+            // Fetch the barcode prefix from POS barcode parser rules.
+            this.weight_barcode_prefix = this.get_barcode_prefix();
+            // Setup the proxy
+            var queue = this.pos.proxy_queue;
+            // The pooling of the scale starts here.
+            var self = this;
+            queue.schedule(function () {
+                return self.pos.proxy.scale_read().then(function (weight) {
+                    try {
+                        self.set_weight(weight);
+                    } catch (error) {
```

Comment:
could the syntax `scale_read().then(function (onFulfilled){...}, function (onRejected){...})` be an alternative for UOM management? `scale_read` in `point_of_sale.devices` uses it

Reply:
I do not get the rationale here, could you please explain it to me ?

Reply:
On a first opinion, I was thinking it would be better to move the UOM management in an overridden `scale_read` function, but my second opinion is that it's better to handle this on the `TareScreenWidget` and not in the `ProxyDevice`